ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt

```diff
@@ -90,14 +108,54 @@ class HomeScreenMapContainerFragment : AbstractMapContainerFragment() {
     map.featureClicks.launchWhenStartedAndCollect { mapContainerViewModel.onFeatureClicked(it) }
   }
 
+  private fun hasValidTasks(cardUiData: MapCardUiData) =
+    when (cardUiData) {
+      // LOI tasks are filtered out of the tasks list for pre-defined tasks.
+      is MapCardUiData.LoiCardUiData ->
+        cardUiData.loi.job.tasks.values.count { !it.isAddLoiTask } > 0
+      is MapCardUiData.AddLoiCardUiData -> cardUiData.job.tasks.values.isNotEmpty()
+    }
+
   /** Invoked when user clicks on the map cards to collect data. */
-  private fun onCollectData(canUserSubmitData: Boolean, cardUiData: MapCardUiData) {
-    if (canUserSubmitData) {
-      navigateToDataCollectionFragment(cardUiData)
-    } else {
+  private suspend fun onCollectData(
+    canUserSubmitData: Boolean,
+    hasTasks: Boolean,
+    hasDataSharingTerms: Survey.DataSharingTerms?,
+    cardUiData: MapCardUiData,
+  ) {
+    if (!canUserSubmitData) {
       // Skip data collection screen if the user can't submit any data
       // TODO(#1667): Revisit UX for displaying view only mode
       ephemeralPopups.ErrorPopup().show(getString(R.string.collect_data_viewer_error))
+      return
+    }
+    if (!hasTasks) {
+      // NOTE(#2539): The DataCollectionFragment will crash if there are no tasks.
+      ephemeralPopups.ErrorPopup().show(getString(R.string.no_tasks_error))
+      return
+    }
+    if (hasDataSharingTerms != null) {
+      (view as ViewGroup).addView(
+        ComposeView(requireContext()).apply {
```

Comment:
Can we move the Compose bit into its own function to avoid excessive nesting here ?